subprojects/scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompiler.java

```diff
@@ -197,29 +211,35 @@ public DefinesClass definesClass(File classpathEntry) {
     private static class ExternalBinariesLookup implements ExternalLookup {
 
         @Override
-        public Option<Changes<File>> changedSources(CompileAnalysis previousAnalysis) {
+        public Option<AnalyzedClass> lookupAnalyzedClass(String binaryClassName, Option<VirtualFileRef> file)  {
+            return Option.empty();
+        }
+
+        @Override
+        public Option<Changes<VirtualFileRef>> changedSources(CompileAnalysis previousAnalysis) {
             return Option.empty();
         }
 
         @Override
-        public Option<Set<File>> changedBinaries(CompileAnalysis previousAnalysis) {
-            List<File> result = new java.util.ArrayList<>();
+        public Option<Set<VirtualFileRef>> changedBinaries(CompileAnalysis previousAnalysis) {
+            List<VirtualFileRef> result = new java.util.ArrayList<>();
 
-            for (Map.Entry<File, Stamp> e : previousAnalysis.readStamps().getAllBinaryStamps().entrySet()) {
-                if (!e.getKey().exists() || !e.getValue().equals(Stamper.forLastModified().apply(e.getKey()))) {
+            for (Map.Entry<VirtualFileRef, Stamp> e : previousAnalysis.readStamps().getAllLibraryStamps().entrySet()) {
+                File path = CONVERTER.toPath(e.getKey()).toFile();
+                if (!path.exists() || !e.getValue().equals(Stamper.forLastModifiedInRootPaths(CONVERTER).apply(e.getKey()))) {
```

Comment:
Would like to raise attention to this line which I think is incorrect and seems to have caused https://github.com/gradle/gradle/issues/20101
e.getValue seems to be a `FarmHash`, not a `LastModified`.

Reply:
TBH I'm not even sure why Gradle provides the `ExternalBinariesLookup` when it effectively could rely on zinc's change detection in the getOrElse case here: https://github.com/sbt/zinc/blob/develop/internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala#L407